### PR TITLE
Solution for reservations on the non-default role

### PR DESCRIPTION
**Pull request by timcharper**
JIRA Issues MARATHON-8662

### CODE REVIEW

**Comment by jeschkies** [comment]
I'm a little confused. The JIRA linked says `Marathon (Kamon) reports metrics to StatsD with default configuration` 🙂

**Comment by jeschkies** [comment]
I'm sorry. I cannot really judge this PR. What is it about?

**Comment by timcharper** [comment]
@jeschkies whoops, the JIRA ticket is 8662, not 8622. I've updated the description.

https://jira.mesosphere.com/browse/MARATHON-8662

**Comment by ANeumann82** [comment]
> Hm, the JIRA says
> 
> ```
> Given a fresh instance of Marathon with the flag --mesos_role value of "*"
> When I post a new resident task /mysql with role "*"
> Then I should receive a validation error response indicating that resident tasks may not have the role "*"
> ```
> 
> is there such a validation already?

Yes, validation is now implemented.

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala [hunk 1]
```diff
@@ -249,6 +249,11 @@ trait PodsValidation extends GeneralPurposeCombinators {
 
   def validPodDefinitionWithRoleEnforcement(roleEnforcement: RoleSettings): Validator[PodDefinition] = validator[PodDefinition] { pod =>
     pod.role is in(roleEnforcement.validRoles)
+    if (pod.isResident) {
+      pod.role is isTrue(s"Resident apps can not have the role ${ResourceRole.Unreserved}") { role: String =>
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]:
```suggestion
      pod.role is isTrue(s"Resident pods cannot have the role ${ResourceRole.Unreserved}") { role: String =>
```

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala [hunk 2]
```diff
@@ -505,6 +505,11 @@ object AppDefinition extends GeneralPurposeCombinators {
 
   def validWithRoleEnforcement(roleEnforcement: RoleSettings): Validator[AppDefinition] = validator[AppDefinition] { app =>
     app.role is in(roleEnforcement.validRoles)
+    if (app.isResident) {
+      app.role is isTrue(s"Resident apps can not have the role ${ResourceRole.Unreserved}") { role: String =>
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]:
```suggestion
      app.role is isTrue(s"Resident apps cannot have the role ${ResourceRole.Unreserved}") { role: String =>
```

File: tests/integration/src/test/scala/mesosphere/marathon/integration/MultiRoleIntegrationTest.scala [hunk 3]
```diff
@@ -38,5 +41,29 @@ class MultiRoleIntegrationTest extends AkkaIntegrationTest with EmbeddedMarathon
...
+//    "Marathon should decline an resident app without a role if mesos_role is *" in {
+//      Given("an app in role dev")
+//      val appInDev = residentApp(PathId("/dev/resident-app-without-role"))
+//
+//      When("The app is created")
+//      val resultInDev = marathon.createAppV2(appInDev)
+//
+//      Then("The apps is not created")
+//      resultInDev should be(Created)
+//    }
```
[the remark this answers is not in the document]
Comment (ANeumann82) [reply tests/integration/src/test/scala/mesosphere/marathon/integration/MultiRoleIntegrationTest.scala]: Yes

**Pull request merged**